### PR TITLE
ENH: Add support for sparse design matrices

### LINKED ISSUE #41: ENH : support sparse data and centering

### PR DESCRIPTION
**Pull request by QB3**
closes #41

### CODE REVIEW

File: andersoncd/datafits.py [hunk 1]
```diff
@@ -24,4 +24,8 @@ def gradient(self, X, y, w, Xw):
         return (X.T @ Xw - self.Xty) / len(Xw)
 
     def gradient_scalar(self, X, w, Xw, j):
+        # TODO remove w unused in argument
```
Comment (mathurinm) [inline andersoncd/datafits.py:27]: We could keep it for datafits where it is present, if it does not hurt ?

File: andersoncd/datafits.py [hunk 2]
```diff
@@ -24,4 +24,8 @@ def gradient(self, X, y, w, Xw):
...
         return (X[:, j] @ Xw - self.Xty[j]) / len(Xw)
+
+    def gradient_scalar_sparse(self, Xj, idx_nz, Xw, y):
+        return (Xj @ (Xw[idx_nz] - y[idx_nz])) / len(Xw)
```
Comment (mathurinm) [inline andersoncd/datafits.py:31]: is this not inefficient as you build the array Xw[...] ? rather than looping for startptr to endptr

Comment (mathurinm) [reply andersoncd/datafits.py:31]: also use Xty

File: andersoncd/solver.py [hunk 3]
```diff
@@ -85,8 +85,8 @@ def solver_path(X, y, datafit, penalty, eps=1e-3, n_alphas=100, alphas=None,
         Maximum violation of KKT along the path.
     """
 
-    if sparse.issparse(X):
-        raise ValueError("Spare design matrices are not supported yet.")
+    # if sparse.issparse(X):
```
Comment (mathurinm) [inline andersoncd/solver.py:88]: rm

File: andersoncd/tests/test_estimators.py [hunk 4]
```diff
@@ -5,12 +5,16 @@
...
 
+from scipy.sparse import csc_matrix
+
 from andersoncd.data import make_correlated_data
 from andersoncd.estimators import Lasso, WeightedLasso, ElasticNet, MCP
 
 X, y, _ = make_correlated_data(
     n_samples=500, n_features=1000, density=0.1, random_state=0)
 
+X_sparse = csc_matrix(X)
```
Comment (mathurinm) [inline andersoncd/tests/test_estimators.py:16]: maybe decimate it before to have a really sparse matrix

**Comment by mathurinm** [comment]
wdyt of having a single kkt_violation, using default arguments, to which you'd alternately pass X or the triplet (data, indices, indptr) as:
```python
def kkt_violation(is_sparse, ..., X=np.zeros(1), data=np.zeros(1), ):
    # stuff 
    if is_sparse:
        # stuff
```

File: andersoncd/solver.py [hunk 5]
```diff
@@ -85,15 +85,15 @@ def solver_path(X, y, datafit, penalty, eps=1e-3, n_alphas=100, alphas=None,
         Maximum violation of KKT along the path.
     """
 
-    if sparse.issparse(X):
-        raise ValueError("Spare design matrices are not supported yet.")
-
     X = check_array(X, 'csc', dtype=[np.float64, np.float32],
```
Comment (mathurinm) [inline andersoncd/solver.py:88]: I'm not sure we accept float32 (given the specs in the numba classes)

**Comment by QB3** [comment]
> wdyt of having a single kkt_violation, using default arguments, to which you'd alternately pass X or the triplet (data, indices, indptr)

I am not sure to understand the benefits of this approach: a function would be removed, but the code quantity would be same no?
I tend to like the consistence of the current approach: all jitted functions that take X in argument are duplicated, with the same signature, exept for X being replaced by X.data, X.intptr, X.....

**Comment by mathurinm** [comment]
For me it reduces code duplication between the two kkt functions

**Comment by mathurinm** [comment]
merge when green @QB3

**Pull request merged**